### ISSUE TITLE
Vimwiki JavaScript library

### ISSUE THREAD

**Issue by chipsenkbeil** [vimwiki/vimwiki#1127]
@tinmarino, @Nudin, Mostly to raise awareness that this is a thing that can potentially be done. Was talking to @brennen on IRC about it. For screenshots of the parser working with an example of syntax highlighting, check out https://github.com/chipsenkbeil/vimwiki-rs/issues/104

Was able to do a proof of concept that takes the vimwiki Rust library, compiles it to wasm using [https://github.com/rustwasm/wasm-bindgen](wasm-bindgen), and then leverages the library to read vimwiki to find headers and replace the headers in a code block with colored span tags.

Potential is both to provide consistent syntax highlighting in browser with the Rust library as well as bring syntax highlighting and other features to editors that leverage typescript (like VS Code).

**Comment by tinmarino** [comment]
Hi @chipsenkbeil,

Thank you for sharing the info here: I had no idea you are in the wasm server.

This is really awesome! I mean the end product: to be able to convert and host my wikis not with Pandoc (or vimwiki hard to maitain converter) + http.server but just in Firefox. Meaning I can put the raw wiki file on github and visit and awsesome html site (if I crafted a awsesome template).

Concerning the implementation, as I said, I did not know about the bind-gen version (even If I knew about the Rust one, I did not think of this possibility).
But I'll have a look as I am interested in learning Rust->Wasm. I already did a little Rust->wasm 3d-hello-world with minimal [html code](https://github.com/tinmarino/tin_wasm/blob/master/index.html) calling bindgen/init which call my [lib.rs:main()](https://github.com/tinmarino/tin_wasm/blob/master/src/lib.rs).

Have you defined an API yet ? The filesystem part can be different in http world. Maybe starting with the data close to the code can help. For example, [there](https://github.com/tinmarino/tin_wasm/blob/357c8deb6d7e9dd1078f402670f928a6fb7eebc8/src/util.rs#L297), the code loads a resource file "res/blue.png" with

```rust
let img = HtmlImageElement::new().unwrap();
imgrc.set_src(img_src);
```

Which is even fewer lines than in JavaScript. This http or filesystem I/O possibility would make your project rocks!

**Comment by chipsenkbeil** [comment]
@tinmarino, here's the different projects I've been working on:

1. **vimwiki-rs**: a Rust library that lets you parse vimwiki documents into Rust structures. I'm currently finishing up the ability to convert those Rust structures into HTML, which means you'll be able to read vimwiki and output to HTML using the Rust library.
2. **vimwiki_macros**: letting you write vimwiki in Rust source code at compile-time. Useful for testing mostly.
3. **vimwiki-server**: a GraphQL server that reads vimwiki documents using the above parser and exposes them via GraphQL so you can query different parts of a document (or multiple documents) such as which pages have the same tag.
4. **vimwiki-wasm**: this was the experiment I was trying. It compiles **vimwiki-rs** into JavaScript using web assembly. I was testing it in that earlier issue and it had no problem parsing vimwiki in the browser and letting me do stuff to it. So, think of this as a JavaScript library version of **vimwiki-rs** that you can use in the browser, in NodeJS, or even in other editors like VS Code that use Typescript.

Alongside the above, I'm also working to integrate vimwiki into [cobalt.rs](https://github.com/cobalt-org/cobalt.rs), a static site generator. This would be the first instance I know of where a static site generator would support the vimwiki language as an option alongside markdown.

**Comment by chipsenkbeil** [comment]
Also, may be some misunderstanding as I'm not writing an HTML framework or dedicated JavaScript app using bindgen. Just experimenting with porting the Rust library to JavaScript so you could parse vimwiki syntax and do neat stuff like syntax highlighting, inspection, etc. directly in the browser.

**Comment by tinmarino** [comment]
The nice thing would be to be able to convert wiki -> html in Javascript. So if you can compile vimwiki-rs to Js, it is already won.

**Comment by chipsenkbeil** [comment]
One of my early thoughts is that if you can do that with JavaScript and it's the same result as with the Rust library then you could have one of those live previews in the browser as you edit your vimwiki documents, which would be neat.

Anyway, figured I'd share here because it was surprisingly easy to get this working. Really impressed with Rust's `wasm-bindgen` library.

**Comment by chipsenkbeil** [comment]
@tinmarino, just verified that html output (still not finished and fully tested) can work with webassembly. Am able to inject the HTML string produced by the library into the DOM with literally three lines of javascript:

```javascript
// Parse vimwiki string
const obj = parse_vimwiki_str("= my header =");

// Inject HTML into output destination
//
// <div id="vimwiki-output">
//     {OUTPUT GOES HERE}
// </div>
const output = document.getElementById("vimwiki-output");
output.insertAdjacentHTML("afterbegin", obj.to_html_str());
```

**Comment by tinmarino** [comment]
@chipsenkbeil really nice. Iguess parse_vimwiki_str is exported by your server.
I hope you feel like me the "usefullness" of this parse_vimwiki_str. This is very motivating (say the one that do not code it jajajaja).
Don't hesitate to push soon a quickstart and ping when done.
Bu the way, I must maks PR of the 3 lines I changes in doc (faster main with link to sub-Readme).

Cheers.

PS: It is cool to read some issues describing aditions, news: better than people complaining that "it still does not work on master" lol.

**Comment by chipsenkbeil** [comment]
@tinmarino I just finished unit testing the html conversion code and merged everything into master. There's also now a crate for a CLI that you can use to compile vimwiki to html. Haven't done anything new with the wasm crate, though.

**Comment by chipsenkbeil** [comment]
@tinmarino I just released version 0.1.0 of vimwiki-rs and all of the related binaries and libraries. I messed up publishing the npm package for vimwiki-wasm, so it's at version 0.1.1 instead XD

https://github.com/chipsenkbeil/vimwiki-rs/blob/master/vimwiki-wasm/README.md

It's currently published for both webpack and nodejs. I need to revise the example for webpack.

If I'm able to configure `package.json` to support a non-webpack version at the same time as webpack and nodejs, I can release a later version to do that. Just don't have much recent experience with npm.

Anyway, closing this out as a real version has finally been released.
